File: handwritten_ocr.js
// handwritten_ocr.js — JavaScript версия

const Tesseract = require('tesseract.js');
const fs = require('fs');
const path = require('path');

class HandwrittenOCR {
    constructor(lang = 'eng', psm = 6, oem = 3) {
        this.lang = lang;
        this.psm = psm;
        this.oem = oem;
        this.results = [];
    }

    async ocrImage(imagePath) {
        console.log(`📂 Обработка: ${path.basename(imagePath)}`);
        const start = Date.now();

        try {
            const result = await Tesseract.recognize(imagePath, this.lang, {
                logger: (m) => {
                    if (m.status === 'recognizing text') {
                        process.stderr.write(`\r⏳ Прогресс: ${Math.round(m.progress * 100)}%`);
                    }
                }
            });

            process.stderr.write('\n');
            const text = result.data.text.trim();
            const elapsed = (Date.now() - start) / 1000;

            this.printResult(text, elapsed);
            this.saveText(text, imagePath);

            return text;
        } catch (err) {
            console.error(`\x1b[31m❌ Ошибка: ${err.message}\x1b[0m`);
            return null;
        }
    }

    async ocrBatch(dir) {
        const exts = ['.jpg', '.jpeg', '.png', '.bmp', '.tiff', '.webp'];
        const files = fs.readdirSync(dir).filter(f => exts.includes(path.extname(f).toLowerCase()));

        if (files.length === 0) {
            console.log('❌ Нет поддерживаемых изображений.');
            return;
        }

        console.log(`📁 Найдено ${files.length} изображений.`);
        for (const file of files) {
            const filePath = path.join(dir, file);
            console.log(`\n[${files.indexOf(file)+1}/${files.length}] Обработка: ${file}`);
            const text = await this.ocrImage(filePath);
            if (text) {
                this.results.push({ file, text });
            }
        }

        this.saveJSON('ocr_results.json');
    }

    printResult(text, elapsed) {
        console.log(`\n\x1b[32mРезультат:\x1b[0m`);
        console.log('─'.repeat(50));
        console.log(text);
        console.log('─'.repeat(50));

        const words = text.split(/\s+/).filter(w => w).length;
        const chars = text.length;
        console.log(`\x1b[36m📊 Статистика:\x1b[0m`);
        console.log(`  Слов: ${words}`);
        console.log(`  Символов: ${chars}`);
        console.log(`  Время: ${elapsed.toFixed(2)} сек`);
    }

    saveText(text, imagePath) {
        const outputPath = path.basename(imagePath, path.extname(imagePath)) + '.txt';
        fs.writeFileSync(outputPath, text);
        console.log(`💾 Сохранено: ${outputPath}`);
    }

    saveJSON(filename) {
        fs.writeFileSync(filename, JSON.stringify(this.results, null, 2));
        console.log(`💾 Сохранено JSON: ${filename}`);
    }
}

async function main() {
    const args = process.argv.slice(2);
    let input = null;
    let lang = 'eng';
    let batch = false;
    let output = null;

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--lang' || args[i] === '-l') lang = args[++i];
        else if (args[i] === '--batch' || args[i] === '-b') batch = true;
        else if (args[i] === '--output' || args[i] === '-o') output = args[++i];
        else if (!input) input = args[i];
    }

    if (!input) {
        console.log('Usage: node handwritten_ocr.js <image> [--lang eng] [--batch]');
        process.exit(1);
    }

    console.log('\x1b[36m✍️ OCR Handwritten Reader (JavaScript)\x1b[0m');

    const ocr = new HandwrittenOCR(lang);

    if (batch || fs.statSync(input).isDirectory()) {
        await ocr.ocrBatch(input);
    } else {
        await ocr.ocrImage(input);
    }
}

main().catch(console.error);

File: handwritten_ocr.rb
# handwritten_ocr.rb — Ruby версия

require 'rtesseract'
require 'json'
require 'optparse'
require 'time'
require 'find'

class HandwrittenOCR
  def initialize(lang = 'eng', psm = 6, oem = 3)
    @lang = lang
    @psm = psm
    @oem = oem
    @results = []
  end

  def ocr_image(image_path)
    puts "📂 Обработка: #{File.basename(image_path)}"
    start = Time.now
    text = RTesseract.new(
      image_path,
      lang: @lang,
      psm: @psm,
      oem: @oem
    ).to_s.strip
    elapsed = Time.now - start

    print_result(text, elapsed)
    save_text(text, image_path)
    text
  end

  def ocr_batch(dir)
    exts = %w[.jpg .jpeg .png .bmp .tiff .webp]
    files = Dir.entries(dir).select { |f| exts.include?(File.extname(f).downcase) }

    if files.empty?
      puts "❌ Нет поддерживаемых изображений."
      return
    end

    puts "📁 Найдено #{files.size} изображений."
    files.each_with_index do |file, i|
      puts "\n[#{i+1}/#{files.size}] Обработка: #{file}"
      text = ocr_image(File.join(dir, file))
      @results << { file: file, text: text } if text
    end

    save_json('ocr_results.json')
  end

  def print_result(text, elapsed)
    puts "\n\e[32mРезультат:\e[0m"
    puts "─" * 50
    puts text
    puts "─" * 50

    words = text.split(/\s+/).reject(&:empty?).size
    chars = text.size
    puts "\e[36m📊 Статистика:\e[0m"
    puts "  Слов: #{words}"
    puts "  Символов: #{chars}"
    puts "  Время: #{elapsed.round(2)} сек"
  end

  def save_text(text, image_path)
    out_file = File.basename(image_path, '.*') + '.txt'
    File.write(out_file, text)
    puts "💾 Сохранено: #{out_file}"
  end

  def save_json(filename)
    File.write(filename, JSON.pretty_generate(@results))
    puts "💾 Сохранено JSON: #{filename}"
  end
end

def main
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: ruby handwritten_ocr.rb <image> [--lang eng] [--batch]"
    opts.on("--lang LANG", "Язык OCR") { |v| options[:lang] = v }
    opts.on("--batch", "Пакетная обработка") { options[:batch] = true }
    opts.on("--output FILE", "Файл для сохранения") { |v| options[:output] = v }
  end.parse!

  input = ARGV[0]
  unless input
    puts "Usage: ruby handwritten_ocr.rb <image> [--lang eng] [--batch]"
    exit 1
  end

  puts "\e[36m✍️ OCR Handwritten Reader (Ruby)\e[0m"

  lang = options[:lang] || 'eng'
  ocr = HandwrittenOCR.new(lang)

  if options[:batch] || File.directory?(input)
    ocr.ocr_batch(input)
  else
    unless File.exist?(input)
      puts "\e[31m❌ Файл не найден: #{input}\e[0m"
      exit 1
    end
    ocr.ocr_image(input)
  end
end

main if __FILE__ == $0
